### ISSUE TITLE
Questions about the result of version committed on Sep 20th,2019

### ISSUE THREAD

**Issue by luyuhao98** [L1aoXingyu/reid_baseline#27]
# Description
Here is my configuration:
```txt
2019-09-24 23:04:53,360 reid_baseline.train INFO: Using 1 GPUs.                                                                                                                                                     
2019-09-24 23:04:53,360 reid_baseline.train INFO: Namespace(config_file='configs/softmax_triplet.yml', opts=['DATASETS.NAMES', '("market1501",)', 'DATASETS.TEST_NAMES', 'market1501', 'SOLVER.IMS_PER_BATCH', '64',
 'MODEL.WITH_IBN', 'True', 'MODEL.BACKBONE', 'resnet50', 'MODEL.PRETRAIN_PATH', '/home/lyh/.cache/torch/checkpoints/resnet50_ibn_a.pth.tar', 'MODEL.VERSION', 'resnet50_ibn_a'])                                    
2019-09-24 23:04:53,360 reid_baseline.train INFO: Loaded configuration file configs/softmax_triplet.yml                                                                                                             
2019-09-24 23:04:53,360 reid_baseline.train INFO: Running with config:                                                                                                                                              
DATALOADER:                                                                                                                                                                                                         
  NUM_INSTANCE: 4                                                                                                                                                                                                   
  NUM_WORKERS: 8                                                                                                                                                                                                    
  SAMPLER: triplet                                                                                                                                                                                                  
DATASETS:                                                                                                                                                                                                           
  NAMES: ('market1501',)                                                                                                                                                                                            
  TEST_NAMES: market1501                                                                                                                                                                                            
INPUT:                                                                                                                                                                                                              
  DO_FLIP: True                                                                                                                                                                                                     
  DO_LIGHTING: False                                                                                                                                                                                                
  DO_PAD: True                                                                                                                                                                                                      
  DO_RE: True                                                                                                                                                                                                       
  FLIP_PROB: 0.5                                                                                                                                                                                                    
  MAX_LIGHTING: 0.2                                                                                                                                                                                                 
  PADDING: 10                                                                                                                                                                                                       
  PADDING_MODE: constant                                                                                                                                                                                            
  PIXEL_MEAN: [0.485, 0.456, 0.406]                                                                                                                                                                                 
  PIXEL_STD: [0.229, 0.224, 0.225]                                                                                                                                                                                  
  P_LIGHTING: 0.75                                                                                                                                                                                                  
  RE_PROB: 0.5                                                                                                                                                                                                      
  SIZE_TEST: [256, 128]                                                                                                                                                                                             
  SIZE_TRAIN: [256, 128]                                                                                                                                                                                            
MODEL:                                                                                                                                                                                                              
  BACKBONE: resnet50                                                                                                                                                                                                
  CHECKPOINT:                                                                                                                                                                                                       
  DIST_BACKEND: dp                                                                                                                                                                                                  
  GCB:                                                                                                                                                                                                              
    ratio: 0.0625                                                                                                                                                                                                   
  LAST_STRIDE: 1                                                                                                                                                                                                                  [35/2062]
  NAME: baseline                                                                                                                                                                                                    
  PRETRAIN: True                                                                                                                                                                                                    
  PRETRAIN_PATH: /home/lyh/.cache/torch/checkpoints/resnet50_ibn_a.pth.tar                                                                                                                                          
  STAGE_WITH_GCB: (False, False, False, False)                                                                                                                                                                      
  VERSION: resnet50_ibn_a                                                                                                                                                                                           
  WITH_IBN: True
OUTPUT_DIR: logs/
SOLVER:
  BASE_LR: 0.00035
  BIAS_LR_FACTOR: 1
  DIST: False
  EVAL_PERIOD: 30
  GAMMA: 0.1
  IMS_PER_BATCH: 64
  LOG_INTERVAL: 30
  LOSSTYPE: ('softmax', 'triplet')
  MARGIN: 0.3
  MAX_EPOCHS: 120
  MOMENTUM: 0.9
  OPT: adam
  STEPS: (40, 90)
  WARMUP_FACTOR: 0.01
  WARMUP_ITERS: 10
  WARMUP_METHOD: linear
  WEIGHT_DECAY: 0.0005
  WEIGHT_DECAY_BIAS: 0.0005
TEST:
  IMS_PER_BATCH: 512
  NORM: True
  WEIGHT: path
```
I ran twice, however, the results are both depressive. 
```txt
Fisrt one:
reid_baseline.train INFO: mAP: 79.3%
reid_baseline.train INFO: CMC curve, Rank-1  :92.2%
reid_baseline.train INFO: CMC curve, Rank-5  :97.2%
reid_baseline.train INFO: CMC curve, Rank-10 :98.2%

Second one:
reid_baseline.train INFO: mAP: 78.9%
reid_baseline.train INFO: CMC curve, Rank-1  :91.7%
reid_baseline.train INFO: CMC curve, Rank-5  :96.9%
reid_baseline.train INFO: CMC curve, Rank-10 :98.1%
```

The script I executed:
```shell
GPUS='0,1'

CUDA_VISIBLE_DEVICES=$GPUS python tools/train.py -cfg='configs/softmax_triplet.yml' \
DATASETS.NAMES '("market1501",)'  \
DATASETS.TEST_NAMES 'market1501' \
SOLVER.IMS_PER_BATCH '64' \
MODEL.WITH_IBN 'True' \
MODEL.BACKBONE 'resnet50' \
MODEL.PRETRAIN_PATH '/home/lyh/.cache/torch/checkpoints/resnet50_ibn_a.pth.tar' \
MODEL.VERSION 'resnet50_ibn_a' \
```

What's more, it seems that in this version, I could not directly utilize **label smooth**  and **gcnet**  just by editing the bash scripts? 

# Questions
1. I changed no codes, so was there anything wrong with my script?
2. If no fatal errors in my script, could you help me with some suggestions to make the result as good as you proposed?  92%(79%) Rank1 on Market1501 is much lower than the expected one, 94%(81%).  It would be great if you would like to show your valid script~
3. As I mentioned above, it seems no direct way to use label_smooth and gcnet that are necessary to get the best result (95% on Market1501). Do you have any good idea if I want to reach it with the least modification?

**Comment by luyuhao98** [comment]
Lol，after I switched my batch from 256 to 64, I got the expected results! Q1 and Q2 don't matter any more~. Maybe too large batch size has a negative effect on hard triple loss. I guess because K-1 of the samples don't provide grad, the larger K causes the larger waste.

**Comment by L1aoXingyu** [comment]
Actual batch size and learning rate are very important for small scale dataset, such as market1501, and some conclusions are not essential. If you train the model on a large scale dataset, you will get a more solid conclusion, but I can't talk about this more.

For Q3, because this is just a pre-release code, I just upload it a few days before. I will add this feature soon. If you want to modify it yourself, you can just change `loss_fns` in `engine/trainer.py`. For GCNet, you can just add `MODEL.STAGE_WITH_GCB = (False, True, True, True)`

**Comment by luyuhao98** [comment]
Thanks a lot. It's really helpful. You are so responsible
